README.md
# Code review Clippy
Forget lousy comments from your teammates - get professional code reviews from Clippy!

![Clippy hates this](ex.png)

## Features

- Passive-aggressive code reviews.

## Usage

- Add api-key
- Select text on the editor 
- Select "Clippy hates your code" from right-click context menu or run directly
## Extension Settings

- Required:
```
{
    "clippyReview.openAiApiKey": "your-api-key"
}
```

- Optional:
```
{
    "clippyReview.openAiApiMaxTokens": 1024,
    "clippyReview.openAiModel": "text-davinci-003"
}
```


CHANGELOG.md
# Change Log

All notable changes to the "code-review-clippy" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/style.ts
export const style = `
body {
	background: black;
    width: 90%;
}

.container {
    display: flex;
    flex-flow: column;
    justify-items: flex-end;
    position: relative;
    top: 2em;
    left: 2em;
    right: 2em;
    bottom: 2em;
    min-height: 40%;
}

.message {
    font-size: 2em;
    position: relative;
    background: #feffc6;
    width: 100%;
    max-width: fit-content;
    min-width: 60vw;
    border-radius: 1em;
    color: black;
    padding: 2em;
    box-shadow: 0px 2px 2px 3px #dadada;
    white-space: pre-line;
}

.message:after {
    content:'';
    position: absolute;
    top: 100%;
    left: 75%;
    margin-left: -50px;
    width: 0;
    height: 0;
    border-top: solid 50px #feffc6;
    border-left: solid 50px transparent;
    border-right: solid 50px transparent;
}

.image {
    max-height: 50%;
    margin-top: 10em;
    margin-left: auto;
}
`;

src/openAiClient.ts
/* eslint-disable @typescript-eslint/naming-convention */
import axios from 'axios';

export interface ClippyOpts {
    apiKey: string;
    model: string;
    maxTokens: number;
    codeToReview: string;
}

const PROMPT_PREFIX = "Write code review for the following code, be as sarcastic and passive-aggressive you can: ";
const COMPLETIONS_ENDPOINT = 'https://api.openai.com/v1/completions';

const settings = {
    temperature: 0.7,
    top_p: 1,
    frequency_penalty: 0,
    presence_penalty: 0,
};

export const getReviewForCode = async (opts: ClippyOpts) => {
    const { apiKey, maxTokens, codeToReview, model } = opts;
    const client = axios.create({
        headers: {
            'Authorization': `Bearer ${apiKey}`,
            'Content-Type': 'application/json',
        }
    });
    const body = {
        ...settings,
        model,
        max_tokens: maxTokens,
        prompt: `${PROMPT_PREFIX} ${codeToReview}`
    };
    return client.post(COMPLETIONS_ENDPOINT, body);
};


src/extension.ts
import path = require('path');
import { style } from './style';
import { getReviewForCode } from './openAiClient';
import * as vscode from 'vscode';

const WORKSPACE = 'clippyReview';
const API_KEY = 'openAiApiKey';
const MODEL = 'openAiModel';
const MAX_TOKENS = 'maxTokens';

const renderHtml = (msg: string, img: vscode.Uri) => `
  <style>
    ${style}
  </style>
  <body>
    <div class="container">
      <div class="message">${msg}</div>
      <img class="image" src="${img}" />
    </div>
  </body>
`;

export function activate(context: vscode.ExtensionContext) {
  let clippy: vscode.WebviewPanel | undefined = undefined;
  let disposable = vscode.commands.registerCommand('code-review-clippy.clippyReview', async () => {
    if (!clippy) {
      clippy = vscode.window.createWebviewPanel("clippy", "Clippy hates your code", {
        viewColumn: vscode.ViewColumn.Beside,
        preserveFocus: true
      });
    }

    const imageUrl = vscode.Uri.file(
      path.join(context.extensionPath, "clippy.png")
    );

    const editor = vscode.window.activeTextEditor;
    let highlight;
    if (editor) {
      let cursorStart = editor.selection.start;
      let cursorEnd = editor.selection.end;
      const range: vscode.Range = new vscode.Range(cursorStart, cursorEnd);
      highlight = editor.document.getText(range);
    }

    const content = await validateAndCreateContent(editor, highlight);
    const img = clippy.webview.asWebviewUri(imageUrl);
    clippy.webview.html = renderHtml(content, img);

    clippy.onDidDispose(
      () => {
        clippy = undefined;
      },
      null,
      context.subscriptions
    );
  });

  context.subscriptions.push(disposable);
}

const validateAndCreateContent = async (editor: vscode.TextEditor | undefined, highlight: string | undefined) => {
  const apiKey = vscode.workspace.getConfiguration(WORKSPACE).get(API_KEY, undefined);
  if (apiKey === null || !apiKey) {
    return toErrorMessage('No OpenAI API-key set!');
  }
  if (!editor) {
    return toErrorMessage('There\'s no active editors open.');
  }
  if (!highlight) {
    return toErrorMessage('There\'s no text on the editor!');
  }
  if (highlight.length > 1000) {
    return toErrorMessage('That\'s a long text, Clippy can\t read that much.');
  }

  const maxTokens = vscode.workspace.getConfiguration(WORKSPACE).get(MAX_TOKENS, 1024);
  const model = vscode.workspace.getConfiguration(WORKSPACE).get(MODEL, 'text-davinci-003');
  const clippyOpts = {
    apiKey: String(apiKey),
    model,
    maxTokens,
    codeToReview: highlight,
  };

  return getReviewForCode(clippyOpts)
    .then((res) => res.data.choices[0].text)
    .catch((e) => toErrorMessage(e.message));
};

const toErrorMessage = (errorMsg: string): string => {
  return `${generateErrorPrefix()}\n\n${errorMsg}`;
};

// These were generated by chatgpt, that's why they suck
const generateErrorPrefix = () => {
  const messages = [
    "Well, this is awkward. It seems like something went wrong. Let's pretend it didn't happen and start over, shall we?",
    "Oopsie daisy! Looks like there's a hiccup in the system. No worries, we'll have it fixed in a jiffy.",
    "Houston, we have a problem. But don't worry, we've got our finest scientists on the case to sort this out.",
    "Whoopsie, did we goof up? It's okay, we all make mistakes. Let's just learn from it and move forward.",
    "Oh dear, it seems like things are not going as planned. Don't worry, we'll get this sorted out and have you on your way in no time.",
    "That's not good.",
    "Clippy doesn\'t like this.",
  ];
  return messages[Math.floor(Math.random() * messages.length)];
};

export function deactivate() { }
